### DISCUSSION TITLE
How to handle failed transactions?

### DISCUSSION THREAD

**Discussion by DreamingInfraviolet** [hyochan/react-native-iap#1561] [Q&A]
Hi!

So it seems that when a purchase succeeds, one has to call `RNIAP.finishTransaction` to tell Apple/Google that the transaction has been successfully delivered.

However, sometimes there may be an error verifying the transaction (network error, fraudulent transaction, etc.). What should one do if there's some kind of error delivering the purchase? It seems that Apple recommends calling `finishTransaction` anyway, but marking the transaction as failed, however I don't see a way of doing this using RNIAP.

Example code illustrating the problem:

```javascript
purchaseUpdatedListener(async (purchase) => {
  try {
    // This can throw an exception verifying/delivering the purchase
    await myServer.deliverProductToUser(purchase);
    await RNIAP.finishTransaction(purchase, false);
  } catch (error) {
    // It's not clear how to handle errors. Should we call RNIAP.finishTransaction, or do something else?
  }
});
```

Right now I can only see RNIAP offering two options when it comes to dealing with errors:

1. If we get an error, call `RNIAP.finishTransaction` anyway. This means that the user will be charged despite not receiving anything, but will correctly clear the ios transaction queue.
2. If we get an error, don't call `RNIAP.finishTransaction`. On Android, the user will be automatically refunded in a few days. On iOS, the behaviour is not clear. The transaction queue is never cleared, which may lead to weird behaviour from what I understand.

Neither of these approaches seem right. From [Apple documentation](https://developer.apple.com/documentation/storekit/original_api_for_in-app_purchase/finishing_a_transaction):

> Your app must finish every transaction, regardless of whether it succeeded or failed.

But this results in users being charged despite not getting anything.

Is there a better way of handling errors?

**Comment by hukpo** [comment]
My suggestion is to finish the transaction ONLY when your receipt is validated successfully on the backend side.
People also suggest finishing it in case SKPaymentTransactionStateFailed happens, but the library already does it for you.
